### PR TITLE
Fix crash when TypedDict contains incorrect metadata

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Fixes at least some of the failures reported in:
https://github.com/jcrist/msgspec/pull/852#issuecomment-2910261218

These were exposed by a bug in 3.14b1 where TypedDict reported
incorrect `__annotations__` but correct `__required_keys__`. msgspec
would crash in this case. The bug is reproducible on earlier Python
versions by manually manipulating attributes on a TypedDict class.

It's a pretty marginal bug but I would argue the extension should
be robust to this sort of edge case.

### CODE REVIEW

File: tests/test_common.py [hunk 1]
```diff
@@ -2099,6 +2099,7 @@ class Base(cls):
 
         class Ex(Base, total=False):
             c: str
+        Ex.__annotations__ = {"c": "str"}
```
Comment (hroncok) [inline tests/test_common.py:2102]: Wouldn't this propagate as RuntimeError to tests that shouldn't have it?

I get e.g.

```
___________ TestTypedDict.test_total_partially_optional[json-False] ____________

self = <test_common.TestTypedDict object at 0x7f88bf0d6620>
proto = <module 'msgspec.json' from '/builddir/build/BUILD/python-msgspec-0.19.0-build/BUILDROOT/usr/lib64/python3.13/site-packages/msgspec/json.py'>
use_typing_extensions = False

    @pytest.mark.parametrize("use_typing_extensions", [False, True])
    def test_total_partially_optional(self, proto, use_typing_extensions):
        if use_typing_extensions:
            tex = pytest.importorskip("typing_extensions")
            cls = tex.TypedDict
        else:
            cls = TypedDict
    
        class Base(cls):
            a: int
            b: str
    
        class Ex(Base, total=False):
            c: str
        Ex.__annotations__ = {"c": "str"}
    
>       dec = proto.Decoder(Ex)

tests/test_common.py:2104: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <class 'test_common.TestTypedDict.test_total_partially_optional.<locals>.Ex'>

    def get_typeddict_info(obj):
        if isinstance(obj, type):
            cls = obj
        else:
            cls = obj.__origin__
    
        raw_hints = get_class_annotations(obj)
    
        if hasattr(cls, "__required_keys__"):
            required = set(cls.__required_keys__)
        elif cls.__total__:
            required = set(raw_hints)
        else:
            required = set()
    
        # Both `typing.TypedDict` and `typing_extensions.TypedDict` have a bug
        # where `Required`/`NotRequired` aren't properly detected at runtime when
        # `__future__.annotations` is enabled, meaning the `__required_keys__`
        # isn't correct. This code block works around this issue by amending the
        # set of required keys as needed, while also stripping off any
        # `Required`/`NotRequired` wrappers.
        hints = {}
        for k, v in raw_hints.items():
            origin = getattr(v, "__origin__", False)
            if origin is Required:
                required.add(k)
                hints[k] = v.__args__[0]
            elif origin is NotRequired:
                required.discard(k)
                hints[k] = v.__args__[0]
            else:
                hints[k] = v
    
        # This can happen if there is a bug in the TypedDict implementation;
        # such a bug was present in Python 3.14.
        if not all(k in hints for k in required):
>           raise RuntimeError(
                f"Required set {required} contains keys that are no in hints: {hints.keys()}"
            )
E           RuntimeError: Required set {'b', 'a'} contains keys that are no in hints: dict_keys(['c'])

../BUILDROOT/usr/lib64/python3.13/site-packages/msgspec/_utils.py:242: RuntimeError
__________ TestTypedDict.test_total_partially_optional[msgpack-False] __________

self = <test_common.TestTypedDict object at 0x7f88bf2e42d0>
proto = <module 'msgspec.msgpack' from '/builddir/build/BUILD/python-msgspec-0.19.0-build/BUILDROOT/usr/lib64/python3.13/site-packages/msgspec/msgpack.py'>
use_typing_extensions = False

    @pytest.mark.parametrize("use_typing_extensions", [False, True])
    def test_total_partially_optional(self, proto, use_typing_extensions):
        if use_typing_extensions:
            tex = pytest.importorskip("typing_extensions")
            cls = tex.TypedDict
        else:
            cls = TypedDict
    
        class Base(cls):
            a: int
            b: str
    
        class Ex(Base, total=False):
            c: str
        Ex.__annotations__ = {"c": "str"}
    
>       dec = proto.Decoder(Ex)

tests/test_common.py:2104: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

obj = <class 'test_common.TestTypedDict.test_total_partially_optional.<locals>.Ex'>

    def get_typeddict_info(obj):
        if isinstance(obj, type):
            cls = obj
        else:
            cls = obj.__origin__
    
        raw_hints = get_class_annotations(obj)
    
        if hasattr(cls, "__required_keys__"):
            required = set(cls.__required_keys__)
        elif cls.__total__:
            required = set(raw_hints)
        else:
            required = set()
    
        # Both `typing.TypedDict` and `typing_extensions.TypedDict` have a bug
        # where `Required`/`NotRequired` aren't properly detected at runtime when
        # `__future__.annotations` is enabled, meaning the `__required_keys__`
        # isn't correct. This code block works around this issue by amending the
        # set of required keys as needed, while also stripping off any
        # `Required`/`NotRequired` wrappers.
        hints = {}
        for k, v in raw_hints.items():
            origin = getattr(v, "__origin__", False)
            if origin is Required:
                required.add(k)
                hints[k] = v.__args__[0]
            elif origin is NotRequired:
                required.discard(k)
                hints[k] = v.__args__[0]
            else:
                hints[k] = v
    
        # This can happen if there is a bug in the TypedDict implementation;
        # such a bug was present in Python 3.14.
        if not all(k in hints for k in required):
>           raise RuntimeError(
                f"Required set {required} contains keys that are no in hints: {hints.keys()}"
            )
E           RuntimeError: Required set {'b', 'a'} contains keys that are no in hints: dict_keys(['c'])

../BUILDROOT/usr/lib64/python3.13/site-packages/msgspec/_utils.py:242: RuntimeError
=========================== short test summary info ============================
FAILED tests/test_common.py::TestTypedDict::test_total_partially_optional[json-False]
FAILED tests/test_common.py::TestTypedDict::test_total_partially_optional[msgpack-False]
```

E.g. with Python 3.13.
